### PR TITLE
bpo-46841: Use *inline* caching for `BINARY_OP`

### PR DESCRIPTION
**Pull request by brandtbucher**
Here's a sort-of-rough first pass at implementing this idea. `BINARY_OP` is easy to start with, since it only uses its cache for the adaptive `counter` member. Initial perf testing doesn't show any impact, which I think is good (it means that the basic idea works well, even when the cache *isn't* being used).

Related changes:
- All of the `dis` utilities grow a `show_caches` parameter, defaulting to `False`.
- `wordcode_helpers.h` has been moved into `compile.c`.

https://bugs.python.org/issue46841

### CODE REVIEW

File: Python/ceval.c [hunk 1]
```diff
@@ -1978,6 +1980,7 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
             if (sub == NULL) {
                 goto error;
             }
+            next_instr += _PyOpcode_InlineCacheEntries[BINARY_OP];
```
Comment (markshannon) [inline Python/ceval.c]: We want to use a macro here. `next_instr += BINARY_OP_CACHE_SIZE`

The reason for this is that we want to define the caches as structs, like [this](https://github.com/faster-cpython/cpython/blob/a20963469638b76f451899bef634fa2e8161fef2/Include/internal/pycore_code.h#L63)

Then we define the `SIZE` macro using `sizeof` for consistency, like [this](https://github.com/faster-cpython/cpython/blob/a20963469638b76f451899bef634fa2e8161fef2/Include/internal/pycore_code.h#L72).

We can then ensure consistency between the struct def and `opcode.py` with an assert:
`assert(BINARY_OP_CACHE_SIZE == _PyOpcode_InlineCacheEntries[BINARY_OP]);`

File: Include/internal/pycore_code.h [hunk 2]
```diff
@@ -132,6 +132,12 @@ _GetSpecializedCacheEntryForInstruction(const _Py_CODEUNIT *first_instr, int nex
     );
 }
 
+static inline uint16_t *
+inline_cache_uint16(_Py_CODEUNIT *next_instr, int offset)
```
Comment (markshannon) [inline Include/internal/pycore_code.h]: It seems like overkill to have an inline function to add its arguments.

File: Python/specialize.c [hunk 3]
```diff
@@ -1874,11 +1878,10 @@ binary_op_fail_kind(int oparg, PyObject *lhs, PyObject *rhs)
 #endif
 
 void
-_Py_Specialize_BinaryOp(PyObject *lhs, PyObject *rhs, _Py_CODEUNIT *instr,
-                        SpecializedCacheEntry *cache)
+_Py_Specialize_BinaryOp(PyObject *lhs, PyObject *rhs, _Py_CODEUNIT *instr)
 {
-    _PyAdaptiveEntry *adaptive = &cache->adaptive;
-    switch (adaptive->original_oparg) {
+    uint16_t *counter = inline_cache_uint16(instr, 1);
```
Comment (markshannon) [inline Python/specialize.c]: `_Py_CODEUNIT *counter = instr+1`

File: Python/ceval.c [hunk 4]
```diff
@@ -5372,23 +5380,23 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
...
             DISPATCH();
         }
 
         TARGET(BINARY_OP_ADAPTIVE) {
             assert(cframe.use_tracing == 0);
-            SpecializedCacheEntry *cache = GET_CACHE();
-            if (cache->adaptive.counter == 0) {
+            uint16_t *counter = inline_cache_uint16(next_instr, 0);
```
Comment (markshannon) [inline Python/ceval.c]: Given that cache entries must be both the same size and alignment as instructions, it makes sense to use `_Py_CODEUNIT` here.

File: Doc/library/dis.rst [hunk 5]
```diff
@@ -24,6 +24,12 @@ interpreter.
       Use 2 bytes for each instruction. Previously the number of bytes varied
       by instruction.
 
+   .. versionchanged:: 3.11
```
Comment (markshannon) [inline Doc/library/dis.rst:27]: Maybe don't document this until we are sure it's what we want?

**Comment by markshannon** [comment]
Looks good in general.
I've commented on https://bugs.python.org/issue46841 about what interface we want to add to the dis module.

File: Doc/library/dis.rst (hunk 5 above)
Comment (brandtbucher) [reply Doc/library/dis.rst:27]: I'd prefer to keep it, and change it if/when `dis` changes (which is easy enough). That way we don't forget to document it.

Comment (markshannon) [reply Doc/library/dis.rst:27]: OK. Once we've started doing this we will need to complete it before the beta release anyway.

**Comment by markshannon** [comment]
:+1:
Let the process of inlining the caches commence

**Pull request merged**